Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -77,7 +77,15 @@ extension HTTPClient {
         }
 
         public init(url: URL, version: HTTPVersion = HTTPVersion(major: 1, minor: 1), method: HTTPMethod = .GET, headers: HTTPHeaders = HTTPHeaders(), body: Body? = nil) throws {
-            guard let scheme = url.scheme else {
+            self.version = version
+            self.method = method
+            self.headers = headers
+            self.body = body
+            try self.setUrl(url)
+        }
+
+        public mutating func setUrl(_ url: URL) throws {
```

Comment:
not sure if the preferred casing is `setUrl` or `setURL` i can go either way

Reply:
I think `setURL` is the preferred casing.

Reply:
Does this need to be `public`? I understand that `redirect()` needs to use it but that doesn't mean it needs to be `public`, does it?

Reply:
I think it should be `public` to make it easier for users to mutate a `Request` they may be holding.

Reply:
OK yes I guess it makes sense given a lot of the rest of the things on `Request` are mutable. We should have doc comments on `public var url/scheme/host` saying that if you want to change the property you should use `setURL()`.

Reply:
imo, mutating the request beyond headers and body is not a common enough use to justify exposing it as a public API, hence the original version of this PR: https://github.com/swift-server/async-http-client/pull/67/commits/88704d319b7280ba2bbd4c647032434d79581b67. the one use case where this is required is redirects, but in those cases you mutate more than just the url so constructing a  new request object is not that bad

@Lukasa and myself went back and forth on this in the context of #56 and i was willing to get that up, for the sake of making progress. 

that said, my opinion has not changed, so would be interesting to get other opinions into the discussion

Reply:
/cc @tanner0101

Reply:
We can go back and forth on this, but I'll state right now at the outset that I think there is no value in the discussion. There is no meaningful difference between a `Request` with a mutable `URL` and one that does not have one as long as it remains a `struct`: literally zero.

Here is the relevant bits of the definition of `Request` from the version of @tomerd's code that was the original, immutable version from this PR:

```swift
public struct Request {
    public let version: HTTPVersion       
    public let method: HTTPMethod
    public let url: URL	
    public var headers: HTTPHeaders
    public var body: Body?

    internal let scheme: String
    internal let host: String

    public init(url: URL, version: HTTPVersion = HTTPVersion(major: 1, minor: 1), method: HTTPMethod = .GET, headers: HTTPHeaders = HTTPHeaders(), body: Body? = nil) throws {
        guard let scheme = url.scheme?.lowercased() else {
	    throw HTTPClientError.emptyScheme
	}
	
        guard Request.isSchemeSupported(scheme: scheme) else {
            throw HTTPClientError.unsupportedScheme(scheme)
        }
        guard let host = url.host else {
            throw HTTPClientError.emptyHost
        }
        self.version = version
        self.method = method
        self.url = url
        self.scheme = scheme
        self.host = host
        self.headers = headers
        self.body = body
    }
}
```

This version, as @tomerd notes, has an immutable `url`, `scheme`, and `host` field. My thesis is that there is nothing you cannot do with this code that you can do with the version we have above, and vice-versa.

Here's my demonstration: I can write the `setURL` function as an extension.

```swift
extension Request {
    mutating func setURL(_ newURL: URL) throws {
        self = try Request(url: newURL, version: self.version, method: self.method, headers: self.headers, body: self.body)
    }
}
```

This function is exactly identical to the version in this patch except that it stresses the optimiser a bit, because it does a bunch of unnecessary work. I can even use it the same way in my code.

My core argument is this: if you have a `struct` there is very little value in making any of its fields immutable when those fields can be set via an initializer, because the setter can always be synthesised by simply calling that initializer. Making a field immutable achieves only two things:

1. Making it awkward to mutate the field
2. Sending a signal that perhaps mutating the field would lead to some logical error.

In this case, neither is true: there is no risk of generating an invalid structure by mutating `url` so long as we use a specific function to achieve it, and users should not be afraid of copying the structure in this way. We don't gain anything by making it hard to do this, and we do push costs onto users by doing it.

Reply:
if i understand you correctly, the difference in opinion boils down to the last statement:

> Sending a signal that perhaps mutating the field would lead to some logical error. .... In this case, neither is true...

the current code has some fragile logic tied to url, scheme and host and mutating the url does cause *logical* issues which is this PR as well as #56 try to address. the fact we are talking about this so much should be a fairly strong signal that there is an issue here. my original  suggestion was to circumvent the current logical issue and potential future ones by not allowing mutation of the url **until we have a good reason to expose it**

> we don't gain anything by making it hard to do this, and we do push costs onto users by doing it

as far as i can judge there is not a good use case to justify exposing this to users behind redirect which we already handle and discussed above, so imo there is no *pragmatic* "cost" here to push onto the users

this is why i still think its a better trade-off to delay exposing a mutable URL until there s a good reason to do so, hope this makes more sense now

Reply:
>  my original suggestion was to circumvent the current logical issue and potential future ones by not allowing mutation of the url until we have a good reason to expose it

And my counter argument is that there is no such thing as "not allowing mutation of the URL": the example code above does so. The question is only whether you can spell it "I'd like to change the URL", or whether you have to spell it "I'd like to create a new Request whose bits all happen to be the same as the one I already have".

I think it's really important to stress that making the `url` property `let` literally does not prevent the users from doing anything that this patch allows them to do. If you can construct a single example of what the current revision of the patch allows that cannot be achieved while the property is `let`, please let me know, because my understanding of Swift is that there is absolutely nothing.

Reply:
@lukasa of course! the point is not that user can't create a new request with a different url. the point is that the internal *logical* state of the request object is easier to manage by making url immutable, since changing the url triggers logic & validation. and since there is not a clear, pragmatic use case to make it mutable, we can get away by making it immutable and save ourselves maintenance headache, at least until a good reason shows up. hope this clarifies where i am coming from

Reply:
worth mentioning again i am okay moving forward with the mutable URL option too, for the sake of making progress. so i suggest at this point we ask @artemredkin @weissi @ianpartridge and @tanner0101 to express their opinions too and just move with whatever the majority feels is the right thing, as both options are technically viable and this is more opinions

Reply:
I think this just represents a fundamental disagreement here, because I don’t think there is any more difficulty in validating the URL via a setter function versus via the constructor. Regardless, though, I think we’re arguing about technicalities: anyone can achieve anything regardless of what construction is used here, so we’ve fundamentally just exerted far more energy here than we need to.

Given that it seems that I’m the only one arguing for a mutator function I think it’s easier for me to simply withdraw the objection than to poll for consensus. Let’s just move on.